### PR TITLE
Audio Requests and Django 1.11

### PR DESCRIPTION
**Pull request by billyeh**
Also upgrade select2

With the usual mix of fixing old bugs, #526, classnotes, and TA leaveslip page layout

### CODE REVIEW

**Comment by billyeh** [comment]
This guy's ready for review before it gets any bigger :) @ksungreborn or @yipingbenjaminliu?

File: ap/audio/views.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  form_class = AudioRequestForm
+  success_url = reverse_lazy('audio:audio-home')
+
+  def form_valid(self, form):
+    req = form.save(commit=False)
+    req.trainee_author = trainee_from_user(self.request.user)
+    req.save()
+    return super(AudioRequestCreate, self).form_valid(form)
+
+class AudioCreate(generic.CreateView):
```
Comment (ksungreborn) [inline ap/audio/views.py]: Should have a regex to ensure the filename follows the standard prescribed to upload audio files

File: ap/audio/templates/audio/ta_buttons.html [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<a href="{% url 'audio:modify-status' status='A' id=request.id %}">
+  <button class="btn btn-success approve" title="Approve"><span class="glyphicon glyphicon-ok"></span></button>
+</a>
+ <a href="{% url 'audio:modify-status' 'D' request.id %}">
+  <button class="btn btn-danger deny" title="Deny"><span class="glyphicon glyphicon-remove"></span></button>
+</a>
+{% if not 'Fellowship' in request.get_status_display %}
+   <a href="{% url 'audio:modify-status' 'F' request.id %}">
+    <button class="btn btn-info fellowship" title="Mark for Fellowship"><span class="glyphicon glyphicon-exclamation-sign"></span></button>
```
Comment (ksungreborn) [inline ap/audio/templates/audio/ta_buttons.html]: Anytime a TA clicks 'Marks for Fellowship' I get error when switching to trainee: 'AudioRequest' object has no attribute 'get_absolute_url'

File: ap/audio/models.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+class AudioRequestManager(models.Manager):
+  def filter_term(self, term):
+    ids = map(lambda f: f.id, AudioFile.objects.filter_term(term))
+    return self.filter(audio_requested__in=ids).distinct()
+
+class AudioRequest(models.Model):
+
+  objects = AudioRequestManager()
+
+  AUDIO_STATUS = (
```
Comment (ksungreborn) [inline ap/audio/models.py]: Should we ensure that only one of any class is present in a single week? I.e. right now I can put multiple BOC 1 in i.e. week 2

File: ap/audio/templates/audio/audiofile_upload.html [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{% extends "base.html" %}
+{% load staticfiles %}
+
+{% block content %}
+  <div class="well">
+    <h3 style="margin-top:0px;">Audio File Upload</h3>
+    Upload files with the name in standard format: B1-01 2017-03-02 DSady.mp3
+    <ul>
+      <li>B1 is the class code</li>
```
Comment (ksungreborn) [inline ap/audio/templates/audio/audiofile_upload.html]: Should the list of class codes be mentioned here for easy reference? Just a consideration

File: ap/audio/views.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    ))
+    return super(AudioHome, self).dispatch(request, *args, **kwargs)
+
+  def get_queryset(self):
+    files = AudioFile.objects.filter_week(self.week)
+    trainee = trainee_from_user(self.request.user)
+    attendance_record = trainee.get_attendance_record()
+    excused_events = filter(lambda r: r['attendance'] == 'E', attendance_record)
+    for f in files:
+      f.request = f.audio_requests.filter(trainee_author=trainee).first()
```
Comment (ksungreborn) [inline ap/audio/views.py]: Can we put all these fields for AudioFile into the model?

File: ap/audio/templates/audio/audiofile_detail.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    <div class="panel-heading">
+      <div class="row">
+        <div class="col-xs-11">
+          {{ audiofile.event.name }} | {{ audiofile.speaker }}
+          <div class="panel-heading-date">{{ audiofile.date }}</div>
+        </div>
+        <div class="col-xs-1">
+          {% if audiofile.request or audiofile.has_leaveslip %}
+            <span class="glyphicon glyphicon-download-alt download-button
+              {% if audiofile.can_download %}download-approved{% endif %}
```
Comment (ksungreborn) [inline ap/audio/templates/audio/audiofile_detail.html]: Maybe we can color-code not just the download-button but also the box of the audio for approved (green), pending (yellow) and denied (red)? Hard to really note the difference in colors of the download-button

File: ap/audio/models.py (hunk 3 above)
Comment (billyeh) [reply ap/audio/models.py]: Nah. I think we should have fewer unnecessary limitations so the code's flexible for the future

File: ap/audio/views.py (hunk 1 above)
Comment (billyeh) [reply ap/audio/views.py]: Added the regex ad33de6 but the class code isn't always a letter and a number.

File: ap/audio/templates/audio/audiofile_detail.html (hunk 6 above)
Comment (billyeh) [reply ap/audio/templates/audio/audiofile_detail.html]: I dunno. I also use the color of the box to indicate which one is playing.

I think we'll have to make more UI changes later so maybe we can table this until then.

File: ap/audio/templates/audio/ta_buttons.html (hunk 2 above)
Comment (billyeh) [reply ap/audio/templates/audio/ta_buttons.html]: That's a good catch. a65e6d5 should fix it.

**Comment by billyeh** [comment]
OK @ksungreborn pending the audio code list I think we're ready to go whenever you want to approve. 😅

File: ap/audio/templates/audio/audiofile_upload.html (hunk 4 above)
Comment (billyeh) [reply ap/audio/templates/audio/audiofile_upload.html]: d99ed83 done.

File: ap/audio/views.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    req.trainee_author = trainee_from_user(self.request.user)
+    req.save()
+    return super(AudioRequestCreate, self).form_valid(form)
+
+class AudioCreate(generic.CreateView):
+  template_name = 'audio/audiofile_upload.html'
+  model = AudioFile
+  fields = []
+
+  def post(self, request):
```
Comment (ksungreborn) [inline ap/audio/views.py]: So when I try and upload a file I get this error now:
TypeError at /audio/upload/ get_available_name() got an unexpected keyword argument 'max_length'.
Tried with title of .mp3 file being "B2-02 2017-09-02 DSady.mp3"

Comment (billyeh) [reply ap/audio/views.py]: Hopefully the new commit fixes this. That was some old code.

Comment (ksungreborn) [reply ap/audio/views.py]: I uploaded with same name and when I try and make an audio download request I get "Event matching query does not exist."

**Pull request merged**